neqo-http3/src/send_message.rs

```diff
@@ -118,33 +104,30 @@ impl SendMessage {
     /// # Errors
     /// `ClosedCriticalStream` if the encoder stream is closed.
     /// `InternalError` if an unexpected error occurred.
-    fn ensure_encoded(&mut self, conn: &mut Connection) -> Res<()> {
-        if let SendMessageState::Initialized { headers, data, fin } = &self.state {
-            qdebug!([self], "Encoding headers");
-            let header_block =
-                self.encoder
-                    .borrow_mut()
-                    .encode_header_block(conn, headers, self.stream_id)?;
-            let hframe = HFrame::Headers {
-                header_block: header_block.to_vec(),
-            };
-            let mut d = Encoder::default();
-            hframe.encode(&mut d);
-            if let Some(buf) = data {
-                qdebug!([self], "Encoding data");
-                let d_frame = HFrame::Data {
-                    len: buf.len() as u64,
-                };
-                d_frame.encode(&mut d);
-                d.encode(buf);
-            }
-
-            self.state = SendMessageState::SendingInitialMessage {
-                buf: d.into(),
-                fin: *fin,
+    fn encode(
+        encoder: &mut QPackEncoder,
+        headers: &[Header],
+        data: Option<&[u8]>,
+        conn: &mut Connection,
+        stream_id: StreamId,
+    ) -> Vec<u8> {
+        qdebug!("Encoding headers");
+        let header_block = encoder.encode_header_block(conn, headers, stream_id);
+        let hframe = HFrame::Headers {
+            header_block: header_block.to_vec(),
+        };
+        let mut d = Encoder::default();
+        hframe.encode(&mut d);
```

Comment:
At some point we should probably try to avoid this bit where we make a vector, put stuff in it, then make another vector and put the other vector in that at an almost fixed offset.  That's unnecessary work.

Our encoder has a neat `encode_vvec_with` function that we could use for encode_header_block so that we encode the frame just once.